raiden_synapse_modules/pfs_presence_router.py

```diff
@@ -0,0 +1,239 @@
+        # Initialise a typed config object
+        config = PFSPresenceRouterConfig()  # type: ignore
+        service_registry_address = config_dict.get("service_registry_address")
+        ethereum_rpc = config_dict.get("ethereum_rpc")
+        blockchain_sync = config_dict.get("blockchain_sync_seconds", "15")
+        try:
+            config.blockchain_sync = int(blockchain_sync)
+        except ValueError:
+            raise ConfigError("`blockchain_sync_seconds` needs to be an integer")
+
+        if service_registry_address is None:
+            raise ConfigError("`service_registry_address` not properly configured")
+        else:
+            try:
+                config.service_registry_address = to_canonical_address(
+                    to_checksum_address(service_registry_address)
+                )
+            except ValueError:
+                raise ConfigError("`service_registry_address` is not a valid address")
+        if ethereum_rpc is None:
+            raise ConfigError("`ethereum_rpc` not properly configured")
+        parsed = urlparse(ethereum_rpc)
+        if not all([parsed.scheme, parsed.netloc]):
+            raise ConfigError("`ethereum_rpc` is not properly configured")
+        else:
+            config.ethereum_rpc = ethereum_rpc
+
+        return config
+
+    async def get_users_for_states(
+        self,
+        state_updates: Iterable[UserPresenceState],
+    ) -> Dict[str, Set[UserPresenceState]]:
+        """Given an iterable of user presence updates, determine where each one
+        needs to go.
+
+        Args:
+            state_updates: An iterable of user presence state updates.
+
+        Returns:
+          A dictionary of user_id -> set of UserPresenceState that the user should
+          receive.
+        """
+        assert state_updates is not None
+        destination_users: Dict[str, Set[UserPresenceState]] = defaultdict(set)
+        for state_update in state_updates:
+            for user in self.local_users:
+                destination_users[user].union({state_update})
+        time_since_last_blockchain_sync = time.time() - self.last_update
+        if time_since_last_blockchain_sync > (2 * self._config.blockchain_sync):
+            log.error(
+                f"Last blockchain sync was {time_since_last_blockchain_sync} seconds"
+                "ago, rescheduling sync."
+            )
+            self._module_api._hs.clock.looping_call(
+                self.check_filters, self._config.blockchain_sync * 1000
+            )
+        return destination_users
+
+    async def get_interested_users(self, user_id: str) -> Union[Set[str], Literal["ALL"]]:
+        """
+        Retrieve a list of users that `user_id` is interested in receiving the
+        presence of. This will be in addition to those they share a room with.
+
+        Optionally, the literal str "ALL" can be returned to indicate that this user
+        should receive all incoming local and remote presence updates.
```

Comment:
Ideally we only receive presence updates for ALL LOCAL presence updates. With our current setup there should not be other presence updates but having misbehavior on the server could trigger this